File: native/tantivy_ex/lib/tantivy_ex/space_analysis.ex
defmodule TantivyEx.SpaceAnalysis do
  @moduledoc """
  Space usage analysis and optimization recommendations for TantivyEx indexes.

  This module provides comprehensive analysis of index space usage, including
  segment breakdown, field-level analysis, and optimization recommendations.

  ## Features

  - Detailed space usage analysis by segments and fields
  - Storage breakdown by component type (postings, terms, fast fields, etc.)
  - Comparison between different analysis snapshots
  - Optimization recommendations based on usage patterns
  - Caching for expensive analysis operations

  ## Usage

      # Create a space analysis resource
      {:ok, analyzer} = TantivyEx.SpaceAnalysis.new()

      # Configure analysis settings
      :ok = TantivyEx.SpaceAnalysis.configure(analyzer, %{
        include_file_details: true,
        include_field_breakdown: true,
        cache_results: true,
        cache_ttl_seconds: 300
      })

      # Analyze an index
      {:ok, analysis} = TantivyEx.SpaceAnalysis.analyze_index(analyzer, index, "snapshot_1")

      # Get optimization recommendations
      {:ok, recommendations} = TantivyEx.SpaceAnalysis.get_recommendations(analyzer, "snapshot_1")

      # Compare two analyses
      {:ok, comparison} = TantivyEx.SpaceAnalysis.compare(analyzer, "snapshot_1", "snapshot_2")
  """

  alias TantivyEx.Native

  @type analysis_resource :: reference()
  @type index_resource :: reference()

  @type analysis_config :: %{
          include_file_details: boolean(),
          include_field_breakdown: boolean(),
          cache_results: boolean(),
          cache_ttl_seconds: pos_integer()
        }

  @type space_analysis :: %{
          total_size_bytes: non_neg_integer(),
          segment_count: non_neg_integer(),
          segments: [segment_analysis()],
          field_analysis: %{String.t() => field_space_usage()},
          index_metadata: index_metadata(),
          storage_breakdown: storage_breakdown()
        }

  @type segment_analysis :: %{
          segment_id: String.t(),
          size_bytes: non_neg_integer(),
          doc_count: non_neg_integer(),
          deleted_docs: non_neg_integer(),
          compression_ratio: float(),
          files: [segment_file()]
        }

  @type segment_file :: %{
          file_type: String.t(),
          file_name: String.t(),
          size_bytes: non_neg_integer(),
          percentage_of_segment: float()
        }

  @type field_space_usage :: %{
          field_name: String.t(),
          total_size_bytes: non_neg_integer(),
          indexed_size_bytes: non_neg_integer(),
          stored_size_bytes: non_neg_integer(),
          fast_fields_size_bytes: non_neg_integer(),
          percentage_of_index: float()
        }

  @type index_metadata :: %{
          total_docs: non_neg_integer(),
          deleted_docs: non_neg_integer(),
          schema_size_bytes: non_neg_integer(),
          num_fields: non_neg_integer(),
          index_settings: %{String.t() => String.t()}
        }

  @type storage_breakdown :: %{
          postings: non_neg_integer(),
          term_dictionary: non_neg_integer(),
          fast_fields: non_neg_integer(),
          field_norms: non_neg_integer(),
          stored_fields: non_neg_integer(),
          positions: non_neg_integer(),
          delete_bitset: non_neg_integer(),
          other: non_neg_integer()
        }

  @type recommendation :: %{
          type: String.t(),
          priority: String.t(),
          description: String.t(),
          potential_savings_bytes: non_neg_integer()
        }

  @doc """
  Create a new space analysis resource.

  ## Returns

  - `{:ok, analysis_resource}` - A new analysis resource
  - `{:error, reason}` - If creation fails

  ## Examples

      {:ok, analyzer} = TantivyEx.SpaceAnalysis.new()
  """
  @spec new() :: {:ok, analysis_resource()} | {:error, term()}
  def new do
    case Native.space_analysis_new() do
      resource when is_reference(resource) -> {:ok, resource}
      error -> {:error, error}
    end
  end

  @doc """
  Configure space analysis settings.

  ## Parameters

  - `analysis_resource` - The analysis resource
  - `config` - Configuration map with analysis settings

  ## Configuration Options

  - `:include_file_details` - Include detailed file breakdown (default: true)
  - `:include_field_breakdown` - Include per-field analysis (default: true)
  - `:cache_results` - Cache analysis results (default: true)
  - `:cache_ttl_seconds` - Cache TTL in seconds (default: 300)

  ## Returns

  - `:ok` - If configuration succeeds
  - `{:error, reason}` - If configuration fails

  ## Examples

      :ok = TantivyEx.SpaceAnalysis.configure(analyzer, %{
        include_file_details: true,
        include_field_breakdown: true,
        cache_results: true,
        cache_ttl_seconds: 600
      })
  """
  @spec configure(analysis_resource(), analysis_config()) :: :ok | {:error, term()}
  def configure(analysis_resource, config) do
    include_file_details = Map.get(config, :include_file_details, true)
    include_field_breakdown = Map.get(config, :include_field_breakdown, true)
    cache_results = Map.get(config, :cache_results, true)
    cache_ttl_seconds = Map.get(config, :cache_ttl_seconds, 300)

    case Native.space_analysis_configure(
           analysis_resource,
           include_file_details,
           include_field_breakdown,
           cache_results,
           cache_ttl_seconds
         ) do
      :ok -> :ok
      error -> {:error, error}
    end
  end

  @doc """
  Analyze space usage for an index.

  Performs comprehensive analysis of index space usage including segments,
  fields, and storage breakdown.

  ## Parameters

  - `analysis_resource` - The analysis resource
  - `index_resource` - The index to analyze
  - `analysis_id` - Unique identifier for this analysis

  ## Returns

  - `{:ok, analysis}` - Space analysis results
  - `{:error, reason}` - If analysis fails

  ## Examples

      {:ok, analysis} = TantivyEx.SpaceAnalysis.analyze_index(analyzer, index, "daily_snapshot")
      # Returns detailed space analysis with size and field information
  """
  @spec analyze_index(analysis_resource(), index_resource(), String.t()) ::
          {:ok, space_analysis()} | {:error, term()}
  def analyze_index(analysis_resource, index_resource, analysis_id) when is_binary(analysis_id) do
    case Native.space_analysis_analyze_index(analysis_resource, index_resource, analysis_id) do
      json_string when is_binary(json_string) ->
        case Jason.decode(json_string, keys: :atoms) do
          {:ok, analysis} -> {:ok, analysis}
          {:error, reason} -> {:error, {:json_decode_error, reason}}
        end

      error ->
        {:error, error}
    end
  end

  @doc """
  Get cached analysis results.

  Retrieves previously cached analysis results if available.

  ## Parameters

  - `analysis_resource` - The analysis resource
  - `analysis_id` - The analysis identifier

  ## Returns

  - `{:ok, analysis_summary}` - If cached results are found
  - `{:error, :not_found}` - If no cached results exist
  - `{:error, reason}` - If retrieval fails

  ## Examples

      case TantivyEx.SpaceAnalysis.get_cached(analyzer, "daily_snapshot") do
        {:ok, summary} ->
          IO.puts("Found cached analysis: \#{summary.total_size_bytes} bytes")

        {:error, :not_found} ->
          # Run new analysis
          TantivyEx.SpaceAnalysis.analyze_index(analyzer, index, "daily_snapshot")
      end
  """
  @spec get_cached(analysis_resource(), String.t()) :: {:ok, map()} | {:error, term()}
  def get_cached(analysis_resource, analysis_id) when is_binary(analysis_id) do
    case Native.space_analysis_get_cached(analysis_resource, analysis_id) do
      json_string when is_binary(json_string) ->
        case Jason.decode(json_string, keys: :atoms) do
          {:ok, %{found: true, analysis: analysis}} -> {:ok, analysis}
          {:ok, %{found: false}} -> {:error, :not_found}
          {:error, reason} -> {:error, {:json_decode_error, reason}}
        end

      error ->
        {:error, error}
    end
  end

  @doc """
  Compare space usage between two analyses.

  Provides detailed comparison showing changes in size, segments, and documents.

  ## Parameters

  - `analysis_resource` - The analysis resource
  - `analysis_id_1` - First analysis identifier
  - `analysis_id_2` - Second analysis identifier

  ## Returns

  - `{:ok, comparison}` - Comparison results
  - `{:error, reason}` - If comparison fails

  ## Examples

      {:ok, comparison} = TantivyEx.SpaceAnalysis.compare(analyzer, "before", "after")

      IO.puts("Size change: \#{comparison.comparison.size_difference_bytes} bytes")
      IO.puts("Change %: \#{comparison.comparison.size_change_percentage}%")
  """
  @spec compare(analysis_resource(), String.t(), String.t()) :: {:ok, map()} | {:error, term()}
  def compare(analysis_resource, analysis_id_1, analysis_id_2)
      when is_binary(analysis_id_1) and is_binary(analysis_id_2) do
    case Native.space_analysis_compare(analysis_resource, analysis_id_1, analysis_id_2) do
      json_string when is_binary(json_string) ->
        case Jason.decode(json_string, keys: :atoms) do
          {:ok, comparison} -> {:ok, comparison}
          {:error, reason} -> {:error, {:json_decode_error, reason}}
        end

      error ->
        {:error, error}
    end
  rescue
    e in ArgumentError -> {:error, "Invalid arguments for compare: #{inspect(e)}"}
    e -> {:error, "Failed to compare analyses: #{inspect(e)}"}
  end

  @doc """
  Get optimization recommendations based on analysis.

  Analyzes space usage patterns and provides actionable recommendations
  for optimizing index storage.

  ## Parameters

  - `analysis_resource` - The analysis resource
  - `analysis_id` - The analysis identifier

  ## Returns

  - `{:ok, recommendations}` - List of optimization recommendations
  - `{:error, reason}` - If analysis fails

  ## Recommendation Types

  - `merge_segments` - Reduce segment count through merging
  - `optimize_deletes` - Clean up deleted documents
  - `field_optimization` - Optimize field storage settings

  ## Examples

      {:ok, %{recommendations: recs}} = TantivyEx.SpaceAnalysis.get_recommendations(analyzer, "snapshot")

      for rec <- recs do
        IO.puts("\#{rec.priority}: \#{rec.description}")
        IO.puts("Potential savings: \#{rec.potential_savings_bytes} bytes")
      end
  """
  @spec get_recommendations(analysis_resource(), String.t()) :: {:ok, map()} | {:error, term()}
  def get_recommendations(analysis_resource, analysis_id) when is_binary(analysis_id) do
    case Native.space_analysis_get_recommendations(analysis_resource, analysis_id) do
      json_string when is_binary(json_string) ->
        case Jason.decode(json_string, keys: :atoms) do
          {:ok, recommendations} -> {:ok, recommendations}
          {:error, reason} -> {:error, {:json_decode_error, reason}}
        end

      error ->
        {:error, error}
    end
  rescue
    e in ArgumentError -> {:error, "Invalid arguments for get_recommendations: #{inspect(e)}"}
    e -> {:error, "Failed to get recommendations: #{inspect(e)}"}
  end

  @doc """
  Clear the analysis cache.

  Removes all cached analysis results to free memory.

  ## Parameters

  - `analysis_resource` - The analysis resource

  ## Returns

  - `:ok` - If cache is cleared successfully
  - `{:error, reason}` - If clearing fails

  ## Examples

      :ok = TantivyEx.SpaceAnalysis.clear_cache(analyzer)
  """
  @spec clear_cache(analysis_resource()) :: :ok | {:error, term()}
  def clear_cache(analysis_resource) do
    case Native.space_analysis_clear_cache(analysis_resource) do
      :ok -> :ok
      error -> {:error, error}
    end
  end

  @doc """
  Get space efficiency metrics.

  Convenience function to extract key efficiency indicators from analysis.

  ## Parameters

  - `analysis` - Space analysis results

  ## Returns

  - `{:ok, metrics}` - Efficiency metrics

  ## Examples

      {:ok, analysis} = TantivyEx.SpaceAnalysis.analyze_index(analyzer, index, "test")
      {:ok, metrics} = TantivyEx.SpaceAnalysis.get_efficiency_metrics(analysis)
      # %{space_per_doc: 1024.5, compression_ratio: 0.75, deletion_ratio: 0.05}
  """
  @spec get_efficiency_metrics(space_analysis()) :: {:ok, map()}
  def get_efficiency_metrics(analysis) do
    space_per_doc =
      if analysis.index_metadata.total_docs > 0 do
        analysis.total_size_bytes / analysis.index_metadata.total_docs
      else
        0.0
      end

    deletion_ratio =
      if analysis.index_metadata.total_docs > 0 do
        analysis.index_metadata.deleted_docs / analysis.index_metadata.total_docs
      else
        0.0
      end

    # Average compression ratio across segments
    avg_compression =
      if analysis.segment_count > 0 do
        analysis.segments
        |> Enum.map(& &1.compression_ratio)
        |> Enum.sum()
        |> Kernel./(analysis.segment_count)
      else
        0.0
      end

    metrics = %{
      space_per_doc: space_per_doc,
      deletion_ratio: deletion_ratio,
      compression_ratio: avg_compression,
      segment_efficiency:
        if analysis.segment_count > 0 do
          analysis.total_size_bytes / analysis.segment_count
        else
          0.0
        end,
      largest_field_percentage:
        analysis.field_analysis
        |> Map.values()
        |> Enum.map(& &1.percentage_of_index)
        |> Enum.max(fn -> 0.0 end)
    }

    {:ok, metrics}
  end

  @doc """
  Format analysis results for human-readable output.

  ## Parameters

  - `analysis` - Space analysis results

  ## Returns

  - `{:ok, formatted_string}` - Human-readable analysis summary

  ## Examples

      {:ok, analysis} = TantivyEx.SpaceAnalysis.analyze_index(analyzer, index, "test")
      {:ok, summary} = TantivyEx.SpaceAnalysis.format_summary(analysis)
      IO.puts(summary)
  """
  @spec format_summary(space_analysis()) :: {:ok, String.t()}
  def format_summary(analysis) do
    size_mb = analysis.total_size_bytes / 1024 / 1024

    summary = ~s"""
    Index Space Analysis Summary
    ============================

    Total Size: #{:erlang.float_to_binary(size_mb, decimals: 2)} MB (#{analysis.total_size_bytes} bytes)
    Segments: #{analysis.segment_count}
    Total Documents: #{analysis.index_metadata.total_docs}
    Deleted Documents: #{analysis.index_metadata.deleted_docs}
    Fields: #{analysis.index_metadata.num_fields}

    Storage Breakdown:
    - Postings: #{:erlang.float_to_binary(analysis.storage_breakdown.postings / 1024 / 1024, decimals: 2)} MB
    - Term Dictionary: #{:erlang.float_to_binary(analysis.storage_breakdown.term_dictionary / 1024 / 1024, decimals: 2)} MB
    - Fast Fields: #{:erlang.float_to_binary(analysis.storage_breakdown.fast_fields / 1024 / 1024, decimals: 2)} MB
    - Stored Fields: #{:erlang.float_to_binary(analysis.storage_breakdown.stored_fields / 1024 / 1024, decimals: 2)} MB
    - Other: #{:erlang.float_to_binary(analysis.storage_breakdown.other / 1024 / 1024, decimals: 2)} MB

    Largest Fields by Size:
    #{analysis.field_analysis |> Map.values() |> Enum.sort_by(& &1.total_size_bytes, :desc) |> Enum.take(5) |> Enum.map(fn field -> "- #{field.field_name}: #{:erlang.float_to_binary(field.total_size_bytes / 1024 / 1024, decimals: 2)} MB (#{:erlang.float_to_binary(field.percentage_of_index, decimals: 1)}%)" end) |> Enum.join("\n")}
    """

    {:ok, summary}
  end
end

File: native/tantivy_ex/native/tantivy_ex/src/modules/space_analysis.rs
use rustler::{Error, NifResult, ResourceArc};
use serde_json;
use std::collections::{BTreeMap, HashMap};
use std::sync::{Arc, Mutex};
use tantivy::{Index, Segment};

use crate::modules::resources::IndexResource;

/// Resource for managing space usage analysis
#[derive(Clone)]
pub struct SpaceAnalysisResource {
    pub analysis_cache: Arc<Mutex<HashMap<String, SpaceAnalysis>>>,
    pub config: Arc<Mutex<AnalysisConfig>>,
}

/// Comprehensive space analysis for an index
#[derive(Debug, Clone)]
pub struct SpaceAnalysis {
    pub total_size_bytes: u64,
    pub segment_count: usize,
    pub segments: Vec<SegmentAnalysis>,
    pub field_analysis: BTreeMap<String, FieldSpaceUsage>,
    pub index_metadata: IndexMetadata,
    pub storage_breakdown: StorageBreakdown,
}

/// Analysis for individual segments
#[derive(Debug, Clone, serde::Serialize)]
pub struct SegmentAnalysis {
    pub segment_id: String,
    pub size_bytes: u64,
    pub doc_count: u32,
    pub deleted_docs: u32,
    pub compression_ratio: f64,
    pub files: Vec<SegmentFile>,
}

/// File information within a segment
#[derive(Debug, Clone, serde::Serialize)]
pub struct SegmentFile {
    pub file_type: String,
    pub file_name: String,
    pub size_bytes: u64,
    pub percentage_of_segment: f64,
}

/// Space usage per field
#[derive(Debug, Clone, serde::Serialize)]
pub struct FieldSpaceUsage {
    pub field_name: String,
    pub total_size_bytes: u64,
    pub indexed_size_bytes: u64,
    pub stored_size_bytes: u64,
    pub fast_fields_size_bytes: u64,
    pub percentage_of_index: f64,
}

/// Index metadata information
#[derive(Debug, Clone)]
pub struct IndexMetadata {
    pub total_docs: u64,
    pub deleted_docs: u64,
    pub schema_size_bytes: u64,
    pub num_fields: usize,
    pub index_settings: BTreeMap<String, String>,
}

/// Storage breakdown by category
#[derive(Debug, Clone)]
pub struct StorageBreakdown {
    pub postings: u64,
    pub term_dictionary: u64,
    pub fast_fields: u64,
    pub field_norms: u64,
    pub stored_fields: u64,
    pub positions: u64,
    pub delete_bitset: u64,
    pub other: u64,
}

/// Configuration for space analysis
#[derive(Debug, Clone)]
pub struct AnalysisConfig {
    pub include_file_details: bool,
    pub include_field_breakdown: bool,
    pub cache_results: bool,
    pub cache_ttl_seconds: u64,
}

// Safety traits for cross-thread usage
unsafe impl Send for SpaceAnalysisResource {}
unsafe impl Sync for SpaceAnalysisResource {}
impl std::panic::RefUnwindSafe for SpaceAnalysisResource {}
impl std::panic::UnwindSafe for SpaceAnalysisResource {}

impl SpaceAnalysisResource {
    pub fn new() -> Self {
        Self {
            analysis_cache: Arc::new(Mutex::new(HashMap::new())),
            config: Arc::new(Mutex::new(AnalysisConfig::default())),
        }
    }
}

impl Default for AnalysisConfig {
    fn default() -> Self {
        Self {
            include_file_details: true,
            include_field_breakdown: true,
            cache_results: true,
            cache_ttl_seconds: 300, // 5 minutes
        }
    }
}

/// Create a new space analysis resource
#[rustler::nif]
pub fn space_analysis_new() -> NifResult<ResourceArc<SpaceAnalysisResource>> {
    let resource = ResourceArc::new(SpaceAnalysisResource::new());
    Ok(resource)
}

/// Configure space analysis settings
#[rustler::nif]
pub fn space_analysis_configure(
    analysis_resource: ResourceArc<SpaceAnalysisResource>,
    include_file_details: bool,
    include_field_breakdown: bool,
    cache_results: bool,
    cache_ttl_seconds: u64,
) -> NifResult<rustler::types::atom::Atom> {
    let mut config = analysis_resource.config.lock().unwrap();
    config.include_file_details = include_file_details;
    config.include_field_breakdown = include_field_breakdown;
    config.cache_results = cache_results;
    config.cache_ttl_seconds = cache_ttl_seconds;

    Ok(rustler::types::atom::ok())
}

/// Analyze space usage for an index
#[rustler::nif(schedule = "DirtyCpu")]
pub fn space_analysis_analyze_index(
    analysis_resource: ResourceArc<SpaceAnalysisResource>,
    index_resource: ResourceArc<IndexResource>,
    analysis_id: String,
) -> NifResult<String> {
    let config = analysis_resource.config.lock().unwrap().clone();

    // Perform comprehensive space analysis
    let total_size = estimate_index_size(&index_resource.index);
    let segment_count = index_resource
        .index
        .searchable_segments()
        .unwrap_or_default()
        .len();

    // Analyze segments
    let segments = analyze_segments(&index_resource.index, &config)?;

    // Analyze fields
    let field_analysis = if config.include_field_breakdown {
        analyze_fields(&index_resource.index)?
    } else {
        BTreeMap::new()
    };

    // Get index metadata
    let metadata = analyze_index_metadata(&index_resource.index)?;

    // Breakdown storage by category
    let storage_breakdown = analyze_storage_breakdown(&index_resource.index, &segments);

    let analysis = SpaceAnalysis {
        total_size_bytes: total_size,
        segment_count,
        segments,
        field_analysis,
        index_metadata: metadata,
        storage_breakdown,
    };

    // Cache results if configured
    if config.cache_results {
        let mut cache = analysis_resource.analysis_cache.lock().unwrap();
        cache.insert(analysis_id, analysis.clone());
    }

    // Convert to JSON response
    let response = serde_json::json!({
        "total_size_bytes": analysis.total_size_bytes,
        "segment_count": analysis.segment_count,
        "segments": analysis.segments.iter().map(|s| {
            let empty_files = Vec::new();
            let files_ref = if config.include_file_details { &s.files } else { &empty_files };
            serde_json::json!({
                "segment_id": s.segment_id,
                "size_bytes": s.size_bytes,
                "doc_count": s.doc_count,
                "deleted_docs": s.deleted_docs,
                "compression_ratio": s.compression_ratio,
                "files": files_ref
            })
        }).collect::<Vec<_>>(),
        "field_analysis": analysis.field_analysis,
        "index_metadata": {
            "total_docs": analysis.index_metadata.total_docs,
            "deleted_docs": analysis.index_metadata.deleted_docs,
            "schema_size_bytes": analysis.index_metadata.schema_size_bytes,
            "num_fields": analysis.index_metadata.num_fields,
            "index_settings": analysis.index_metadata.index_settings
        },
        "storage_breakdown": {
            "postings": analysis.storage_breakdown.postings,
            "term_dictionary": analysis.storage_breakdown.term_dictionary,
            "fast_fields": analysis.storage_breakdown.fast_fields,
            "field_norms": analysis.storage_breakdown.field_norms,
            "stored_fields": analysis.storage_breakdown.stored_fields,
            "positions": analysis.storage_breakdown.positions,
            "delete_bitset": analysis.storage_breakdown.delete_bitset,
            "other": analysis.storage_breakdown.other
        }
    });

    Ok(response.to_string())
}

/// Get cached analysis results
#[rustler::nif]
pub fn space_analysis_get_cached(
    analysis_resource: ResourceArc<SpaceAnalysisResource>,
    analysis_id: String,
) -> NifResult<String> {
    let cache = analysis_resource.analysis_cache.lock().unwrap();

    if let Some(analysis) = cache.get(&analysis_id) {
        let response = serde_json::json!({
            "found": true,
            "analysis": {
                "total_size_bytes": analysis.total_size_bytes,
                "segment_count": analysis.segment_count,
                "field_count": analysis.field_analysis.len(),
                "total_docs": analysis.index_metadata.total_docs
            }
        });
        Ok(response.to_string())
    } else {
        let response = serde_json::json!({
            "found": false
        });
        Ok(response.to_string())
    }
}

/// Compare space usage between two analyses
#[rustler::nif(schedule = "DirtyCpu")]
pub fn space_analysis_compare(
    analysis_resource: ResourceArc<SpaceAnalysisResource>,
    analysis_id_1: String,
    analysis_id_2: String,
) -> NifResult<String> {
    let cache = analysis_resource.analysis_cache.lock().unwrap();

    let analysis_1 = cache.get(&analysis_id_1).ok_or(Error::BadArg)?;
    let analysis_2 = cache.get(&analysis_id_2).ok_or(Error::BadArg)?;

    let size_diff = analysis_2.total_size_bytes as i64 - analysis_1.total_size_bytes as i64;
    let segment_diff = analysis_2.segment_count as i32 - analysis_1.segment_count as i32;
    let doc_diff =
        analysis_2.index_metadata.total_docs as i64 - analysis_1.index_metadata.total_docs as i64;

    let response = serde_json::json!({
        "comparison": {
            "size_difference_bytes": size_diff,
            "size_change_percentage": if analysis_1.total_size_bytes > 0 {
                (size_diff as f64 / analysis_1.total_size_bytes as f64) * 100.0
            } else {
                0.0
            },
            "segment_difference": segment_diff,
            "document_difference": doc_diff,
            "analysis_1": {
                "id": analysis_id_1,
                "size_bytes": analysis_1.total_size_bytes,
                "segments": analysis_1.segment_count,
                "docs": analysis_1.index_metadata.total_docs
            },
            "analysis_2": {
                "id": analysis_id_2,
                "size_bytes": analysis_2.total_size_bytes,
                "segments": analysis_2.segment_count,
                "docs": analysis_2.index_metadata.total_docs
            }
        }
    });

    Ok(response.to_string())
}

/// Get space optimization recommendations
#[rustler::nif]
pub fn space_analysis_get_recommendations(
    analysis_resource: ResourceArc<SpaceAnalysisResource>,
    analysis_id: String,
) -> NifResult<String> {
    let cache = analysis_resource.analysis_cache.lock().unwrap();
    let analysis = cache.get(&analysis_id).ok_or(Error::BadArg)?;

    let mut recommendations = Vec::new();

    // Check for too many segments
    if analysis.segment_count > 10 {
        recommendations.push(serde_json::json!({
            "type": "merge_segments",
            "priority": "high",
            "description": format!("Consider merging segments. Current count: {}", analysis.segment_count),
            "potential_savings_bytes": estimate_merge_savings(analysis)
        }));
    }

    // Check for deleted documents
    let deletion_ratio =
        analysis.index_metadata.deleted_docs as f64 / analysis.index_metadata.total_docs as f64;
    if deletion_ratio > 0.1 {
        recommendations.push(serde_json::json!({
            "type": "optimize_deletes",
            "priority": "medium",
            "description": format!("High deletion ratio: {:.1}%. Consider optimization.", deletion_ratio * 100.0),
            "potential_savings_bytes": estimate_deletion_savings(analysis)
        }));
    }

    // Check for field storage efficiency
    for (_field_name, field_usage) in &analysis.field_analysis {
        if field_usage.percentage_of_index > 50.0 {
            recommendations.push(serde_json::json!({
                "type": "field_optimization",
                "priority": "low",
                "description": format!("Field '{}' uses {:.1}% of index space", field_usage.field_name, field_usage.percentage_of_index),
                "potential_savings_bytes": 0
            }));
        }
    }

    let response = serde_json::json!({
        "analysis_id": analysis_id,
        "recommendations": recommendations,
        "total_recommendations": recommendations.len()
    });

    Ok(response.to_string())
}

/// Clear analysis cache
#[rustler::nif]
pub fn space_analysis_clear_cache(
    analysis_resource: ResourceArc<SpaceAnalysisResource>,
) -> NifResult<rustler::types::atom::Atom> {
    let mut cache = analysis_resource.analysis_cache.lock().unwrap();
    cache.clear();
    Ok(rustler::types::atom::ok())
}

// Helper functions for space analysis

fn estimate_index_size(_index: &Index) -> u64 {
    // Simplified estimation - in a real implementation, this would walk the directory
    // and sum up all file sizes
    1024 * 1024 * 10 // 10MB placeholder
}

fn analyze_segments(index: &Index, config: &AnalysisConfig) -> NifResult<Vec<SegmentAnalysis>> {
    let mut segments = Vec::new();

    if let Ok(searchable_segments) = index.searchable_segments() {
        for (i, segment) in searchable_segments.iter().enumerate() {
            let segment_analysis = SegmentAnalysis {
                segment_id: format!("segment_{}", i),
                size_bytes: 1024 * 1024, // Placeholder
                doc_count: 1000,         // Placeholder - would need segment reader
                deleted_docs: 0,         // Placeholder - would need segment reader
                compression_ratio: 0.8,  // Placeholder
                files: if config.include_file_details {
                    analyze_segment_files(segment)
                } else {
                    Vec::new()
                },
            };
            segments.push(segment_analysis);
        }
    }

    Ok(segments)
}

fn analyze_segment_files(_segment: &Segment) -> Vec<SegmentFile> {
    // Placeholder implementation
    vec![
        SegmentFile {
            file_type: "postings".to_string(),
            file_name: "postings.idx".to_string(),
            size_bytes: 512 * 1024,
            percentage_of_segment: 50.0,
        },
        SegmentFile {
            file_type: "terms".to_string(),
            file_name: "terms.idx".to_string(),
            size_bytes: 256 * 1024,
            percentage_of_segment: 25.0,
        },
    ]
}

fn analyze_fields(index: &Index) -> NifResult<BTreeMap<String, FieldSpaceUsage>> {
    let mut field_analysis = BTreeMap::new();
    let schema = index.schema();

    for (_field, field_entry) in schema.fields() {
        let field_name = field_entry.name().to_string();
        let usage = FieldSpaceUsage {
            field_name: field_name.clone(),
            total_size_bytes: 1024 * 1024, // Placeholder
            indexed_size_bytes: 512 * 1024,
            stored_size_bytes: 256 * 1024,
            fast_fields_size_bytes: 256 * 1024,
            percentage_of_index: 10.0, // Placeholder
        };
        field_analysis.insert(field_name, usage);
    }

    Ok(field_analysis)
}

fn analyze_index_metadata(index: &Index) -> NifResult<IndexMetadata> {
    let schema = index.schema();
    let reader = index.reader().map_err(|_| Error::BadArg)?;
    let searcher = reader.searcher();

    let metadata = IndexMetadata {
        total_docs: searcher.num_docs() as u64,
        deleted_docs: 0,         // Simplified
        schema_size_bytes: 1024, // Placeholder
        num_fields: schema.fields().count(),
        index_settings: BTreeMap::new(), // Placeholder
    };

    Ok(metadata)
}

fn analyze_storage_breakdown(_index: &Index, segments: &[SegmentAnalysis]) -> StorageBreakdown {
    // Simplified analysis based on segment data
    let total_size: u64 = segments.iter().map(|s| s.size_bytes).sum();

    StorageBreakdown {
        postings: total_size / 3,
        term_dictionary: total_size / 6,
        fast_fields: total_size / 6,
        field_norms: total_size / 12,
        stored_fields: total_size / 6,
        positions: total_size / 12,
        delete_bitset: total_size / 24,
        other: total_size / 24,
    }
}

fn estimate_merge_savings(analysis: &SpaceAnalysis) -> u64 {
    // Estimate savings from merging segments (typically 10-20% for many small segments)
    if analysis.segment_count > 5 {
        analysis.total_size_bytes / 10 // 10% savings estimate
    } else {
        0
    }
}

fn estimate_deletion_savings(analysis: &SpaceAnalysis) -> u64 {
    // Estimate savings from optimizing deleted documents
    let deletion_ratio =
        analysis.index_metadata.deleted_docs as f64 / analysis.index_metadata.total_docs as f64;
    (analysis.total_size_bytes as f64 * deletion_ratio * 0.8) as u64 // 80% of deleted doc space can be reclaimed
}
